refactor(color_input): migrate ColorInput to a function component

Victor-Nyagudi opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
This pull request migrates `ColorInput` from a class component to a function component and migrates its tests from Enzyme to React Testing Library.

[Mattermost conversation on function component migration](https://community.mattermost.com/core/pl/nxfp9dsx9tf9ie86n1in9418wy) for reference.

#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/XXX
Jira https://mattermost.atlassian.net/browse/MM-XXX
-->
N/A

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->
N/A

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
NONE
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Change Impact: 🟢 Low

**Regression Risk:** This is an isolated refactor of a single UI component from class to function component with direct test migration from Enzyme to React Testing Library. The public API remains unchanged, and test coverage has been maintained with expanded user-interaction assertions, minimizing the likelihood of breaking existing behavior.

**QA Recommendation:** Minimal manual QA required. Verify the ColorInput component displays and accepts color values correctly in context where it's used (theme/branding color pickers). Automated test coverage appears comprehensive with React Testing Library assertions covering open/close behavior, input handling, and value normalization.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Victor-Nyagudi committed:
feat(color_input): migrate ColorInput to function component

Victor-Nyagudi committed:
test(color_input): migrate tests from enzyme to react testing library

Victor-Nyagudi committed:
test(color_input): remove obsolete snapshots

Victor-Nyagudi committed:
test(color_input): update snapshots

agnivade commented on webapp/channels/src/components/color_input.tsx:
```diff
@@ -13,172 +13,155 @@ type Props = {
+         * Since 'isOpened' is changed by a click event, this 'useEffect' runs before the screen is painted
+         * therefore, 'checkClick' is fired before the component is mounted which in turn calls 'setIsOpened(false)'.
+         * It's not possible to update the state of an unmounted component. 'setTimeout' ensures the event listener is
+         * added after the screen is painted and the component mounted. The delay (1) is just the smallest number possible
```
How are we sure that 1ms is sufficient for this?

agnivade replied:
Yep, sounds good!

hmhealey commented on webapp/channels/src/components/color_input.tsx:
```diff
@@ -13,172 +13,155 @@ type Props = {
+         * so the listeners are attached as soon as possible.
+         * https://react.dev/reference/react/useEffect#:~:text=If%20your%20Effect%20is%20caused%20by%20an%20interaction%20(like%20a%20click)%2C%20React%20may%20run%20your%20Effect%20before%20the%20browser%20paints%20the%20updated%20screen.
+         * */
+        setTimeout(() => {
```
Could we perhaps get rid of the `setTimeout` this to use a mousedown instead of a click? There's similar logic in the Global Header's `useClickOutsideRef` hook which uses that event, and it seems to work in a more straightforward way.

Alternatively, if you want to go the extra mile here, we could replace all of this if we moved the colour picker to use Floating UI since it handles the click outside for us as well as other stuff like the focus trapping logic. Take a look at the `useEmojiPicker` hook (`webapp/channels/src/components/emoji_picker/use_emoji_picker.tsx`) if you're interested in doing that.

hmhealey replied:
We've migrated everything that used React Bootstrap, Popper.js, and React Popper to Floating UI a while back, but we haven't looked for places like this that implement floating elements themselves

hmhealey replied:
Sorry, I forgot to answer that earlier. I'm fine with leaving the `setTimeout` here. Even if it looks a bit weird, it seems like it should work correctly

hmhealey commented on webapp/channels/src/components/color_input.tsx:
```diff
@@ -13,172 +13,155 @@ type Props = {
+    const [valueFromState, setValueFromState] = useState(valueFromProps);
+
+    if (!isFocused && valueFromProps !== valueFromState) {
+        setValueFromState(valueFromProps);
```
Do we still need this when we do something similar in the `onBlur` handler? I'm not sure if this was added for some case like when you close the picker or something, but it seems odd to be doing this in this way. I don't know if setting state from within the body of a function component would cause an actual problem or if it just looks unusual

mattermost-build committed:
Merge branch 'master' into migrate/color-input-to-function-component

Victor-Nyagudi committed:
test(color_input): update test

Victor-Nyagudi committed:
refactor(color_input): remove unnecessary state update in function body

Victor-Nyagudi commented:
Thought the "review" option would just batch my comments into one notification, but I see it's started a whole new review.

I'll just leave the comments individually going forward to not confusing things.

hmhealey reviewed (changes requested):
Sorry, I reviewed this a couple weeks ago, but apparently I forgot to click submit 😅

Victor-Nyagudi committed:
Revert "refactor(color_input): remove unnecessary state update in function body"

This reverts commit 2c7647a3e49075535f3bcfd810b3d2de8173a7a0.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

Victor-Nyagudi commented:
/update-branch

mattermost-build committed:
Merge branch 'master' into migrate/color-input-to-function-component

hmhealey reviewed (changes requested):
Sorry, I know it's been a while, but coming back to this, I'm trying to remember where we were with it. I know I suggested some changes to make the tests a bit more consistent, but I'm fine to leave those as-is since the current ones are fine enough outside of the one place where they're failing:
```
FAIL src/components/color_input.test.tsx
  ● components/ColorInput › should keep what the user types in the textbox until blur

    expect(element).toHaveValue(#aabbcc)

    Expected the element to have value:
      #aabbcc
    Received:
      #ffffff

      76 |
      77 |         expect(document.activeElement).not.toBe(inputElement);
    > 78 |         expect(inputElement).toHaveValue('#aabbcc');
         |                              ^
      79 |         expect(baseProps.onChange).toHaveBeenLastCalledWith('#aabbcc');
      80 |
      81 |         // The RGB value passed in the assertion is the equivalent of '#aabbcc'.

      at Object.toHaveValue (src/components/color_input.test.tsx:78:30)
```

There's some other failures in the tests, but I hope those are just a side effect of the new test failing

mattermost-build committed:
Merge branch 'master' into migrate/color-input-to-function-component

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

Victor-Nyagudi commented:
@hmhealey I kind of forgot about this PR.

Unless you have any objections, I'm in favor of closing it based on [this thread](https://community.mattermost.com/core/pl/9ewkzfoz7pg4zxao9unnoq96wh).
> In order to streamline efforts, I'll stop migrating the existing class components until a proper plan is in place of how to go about using the aforementioned AI agent. If there's a need for external help with reviewing the AI code, I'm willing to lend a hand.

cc: @agnivade

hmhealey committed:
Merge remote-tracking branch 'origin/master' into migrate/color-input-to-function-component

hmhealey commented:
Let me see if I can get this fixed up. I'd still like to get this in so that we aren't wasting your contribution 🙂

hmhealey committed:
Fix ColorInput tests

hmhealey committed:
Simplify click outside handler

By changing the button to always show the picker instead of toggling it
and making it so that the click outside handler checks for clicks
outside of the whole ColorInput, we can get rid of the setTimeout and
the very specific timing needed for the click outside handler.

hmhealey committed:
Wrap handleColorChange in useCallback

hmhealey committed:
Update snapshot

hmhealey commented:
Finally got CI to behave 😅. Thanks again for the PR, Victor!

Victor-Nyagudi commented:
No problem.
